src/Stratis.Bitcoin/P2P/SelfEndpointTracker.cs

```diff
@@ -22,5 +33,48 @@ public bool IsSelf(IPEndPoint ipEndPoint)
         {
             return this.knownSelfEndpoints.Contains(ipEndPoint);
         }
+
+        /// <summary>Updates external IP address and peer score of the node.</summary>
+        public void UpdateAndAssignMyExternalAddress(IPEndPoint ipEndPoint, ConnectionManager connectionManager)
```

Comment:
Do the parameters need guards ?

Reply:
I am struggling to understand what this method is doing when the summary says it's updating external IP addresses and working out the Peer score of the node.

Reply:
I can see two actions in this method - SetExternalAddress and SetExternalEndPoint.  Maybe move these into separate methods.  I'm not 100% sure if this is required, but something like this - (I haven't run this) ;

```
/// <summary>Updates external IP address and peer score of the node.</summary>
public void UpdateAndAssignMyExternalAddress(IPEndPoint ipEndPoint, ConnectionManager connectionManager)
{
    // If none supplied or able to route, take from version handshake.
    if (ipEndPoint == null)
        return;

    if (this.SetExternalEndpoint(connectionManager)) return;

    connectionManager.ConnectionSettings.ExternalEndpoint = this.SetExternalAddress(ipEndPoint).Key;
}

private KeyValuePair<IPEndPoint, int> SetExternalAddress(IPEndPoint ipEndPoint)
{
    // If it was the same as value that was there we just increment the score by 1.
    if (ipEndPoint.Equals(this.MyExternalAddress.Key))
    {
        this.MyExternalAddress =
            new KeyValuePair<IPEndPoint, int>(this.MyExternalAddress.Key, this.MyExternalAddress.Value + 1);
    }

    // If it was different we decrement the score by 1.
    this.MyExternalAddress =new KeyValuePair<IPEndPoint, int>(this.MyExternalAddress.Key, this.MyExternalAddress.Value - 1);

    // If the new score is 0 we replace the old one with the new one with score 1.
    return this.MyExternalAddress.Value <= 0 ? new KeyValuePair<IPEndPoint, int>(ipEndPoint, 1) : default(KeyValuePair<IPEndPoint, int>);
}

private bool SetExternalEndpoint(IConnectionManager connectionManager)
{
    // if external IP address supplied this overrides all.
    if (connectionManager.ConnectionSettings.ExternalIpFromNodeConfiguration != null)
    {
        connectionManager.ConnectionSettings.ExternalEndpoint =
            connectionManager.ConnectionSettings.ExternalIpFromNodeConfiguration;
        return true;
    }

    // If external IP address not supplied take first routeable bind address and set score to 10.
    IPEndPoint nodeServerEndpoint = connectionManager.ConnectionSettings.Listen
        ?.FirstOrDefault(x => x.Endpoint.Address.IsRoutable(false))?.Endpoint;

    if (nodeServerEndpoint == null) return false;

    this.MyExternalAddress = new KeyValuePair<IPEndPoint, int>(nodeServerEndpoint, 10);
    connectionManager.ConnectionSettings.ExternalEndpoint = this.MyExternalAddress.Key;

    return true;
}
```